### PR TITLE
Add handling for integrity failures when applying and resolving change lists

### PR DESCRIPTION
**Pull request by jonawals**
**UPDATE**

The commit hash in question (`81f9afdf13c2361e09691cd85ae2ef62ed017786`) associated with the seeding of the test impact data _was_ in fact out of sync, instead referring to a commit from 3 days prior to the seed snapshot. This was due to the fact that the TIAF script was incorrectly handling the updating of the last commit hash resulting in a stale hash being used instead of that from which the binary used to seed was built.

***

**Background**

The PR build for PR-1960 caused the TIAF runtime to abort due to the newly created file `Gems\ScriptCanvas\Code\Builder\ScriptCanvasBuilder.h` already having an entry in the source dependency map, implying that a previous change list containing a CRUD delete for this file has been missed, resulting in a compromised source dependency map. 

**Reproduction of the Bug**

1. I checked out commit `81f9afdf13c2361e09691cd85ae2ef62ed017786` where the seeding took place and ran the TIAF driver script with the same args as that of the TIAF job linked above to generate the test impact data in the `Persistent/active` directory and the last commit hash in the `Persistent/historic` directory of the TIAF working directory . 
2. I checked out commit `ea0e38aaf4b58d7142e2033d9daeeef9fd5fb58b` as per the job above and ran the TIAF driver script to construct the change list between these two commits and run the TIAF console. 

I could not re-create the bug above because `Gems\ScriptCanvas\Code\Builder\ScriptCanvasBuilder.h` did not have an entry in the test impact data when Step 1 (above) was run, thus the call to `ApplyAndResoveChangeList` did not fail. As there is currently no way to access the test impact data generated by the seeding on the `development` branch that was subsequently captured in the nightly snapshot I cannot conclusively determine what the cause is. As the aforementioned file does appear to be newly created in commit `ea0e38aaf4b58d7142e2033d9daeeef9fd5fb58b` one possible explanation is that for whatever reason the test impact data was updated at some indeterminate point but the last commit hash was not. 

**Solution**

Immediate: Prior to this PR, such compromises to the integrity of the source dependency were considered hard failures and would throw a `DependencyException`, causing the TIAF runtime to abort. The integrity failure policy used by the TIAF runtime is now also used by the dynamic dependency map for handling integrity failures during `ApplyAndResoveChangeList`, allowing the client to override the failure and force the change list application and resolution to continue. The default policy for  test impact analysis sequences (not seed sequences) is now `continue`.

Future: I will store the last commit hash and test impact data together. This will be achieved by the TIAF driver rolling the last commit hash and the test impact data together in a JSON file upon seeding to be unpacked each time for test impact analysis sequences. This, along with a time stamp and any other pertinent data, will allow the state of the test impact data to be inspected in more detail to allow for a more accurate reproduction of future integrity failure issues.

**Log Output of PR-1960**
```[2021-07-08T14:19:02.305Z] D:\workspace\o3de>python\python.cmd -u scripts\build\ci_build.py --platform Windows --type test_impact_analysis_profile_vs2019 

[2021-07-08T14:19:02.305Z] [ci_build] Executing "D:\workspace\o3de\scripts\build\Platform/Windows/python_windows.cmd"

[2021-07-08T14:19:02.305Z]   cwd = D:\workspace\o3de

[2021-07-08T14:19:02.305Z]   engine_dir = D:\workspace\o3de

[2021-07-08T14:19:02.305Z]   paramaters:

[2021-07-08T14:19:02.305Z]     OUTPUT_DIRECTORY = build/windows_vs2019 

[2021-07-08T14:19:02.305Z]     CONFIGURATION = profile 

[2021-07-08T14:19:02.305Z]     SCRIPT_PATH = scripts/build/TestImpactAnalysis/tiaf_driver.py 

[2021-07-08T14:19:02.305Z]     SCRIPT_PARAMETERS = --config="!OUTPUT_DIRECTORY!/bin/TestImpactFramework/persistent/tiaf.profile.json" --suite=main --test-failure-policy=continue --src-branch=!BRANCH_NAME! --dst-branch=!CHANGE_TARGET! --pipeline=!PIPELINE_NAME! --dest-commit=!CHANGE_ID! --seeding-branches=!BUILD_SNAPSHOTS! --seeding-pipelines=default 

[2021-07-08T14:19:02.305Z] --------------------------------------------------------------------------------

[2021-07-08T14:19:02.305Z] [ci_build] python/python.cmd -u scripts/build/TestImpactAnalysis/tiaf_driver.py --config="build/windows_vs2019/bin/TestImpactFramework/persistent/tiaf.profile.json" --suite=main --test-failure-policy=continue --src-branch=PR-1960 --dst-branch=development --pipeline=default --dest-commit=ea0e38aaf4b58d7142e2033d9daeeef9fd5fb58b --seeding-branches=development,stabilization/2106 --seeding-pipelines=default

[2021-07-08T14:19:04.211Z] Commit: 'ea0e38aaf4b58d7142e2033d9daeeef9fd5fb58b'.

[2021-07-08T14:19:04.211Z] Source branch: 'PR-1960'.

[2021-07-08T14:19:04.211Z] Destination branch: 'development'.

[2021-07-08T14:19:04.211Z] Seeding branches: '['development', 'stabilization/2106']'.

[2021-07-08T14:19:04.211Z] Is seeding branch: 'False'.

[2021-07-08T14:19:04.211Z] Pipeline: 'default'.

[2021-07-08T14:19:04.211Z] Seeding pipelines: '['default']'.

[2021-07-08T14:19:04.211Z] Is seeding pipeline: 'True'.

[2021-07-08T14:19:04.211Z] Attempting to parse configuration file 'build/windows_vs2019/bin/TestImpactFramework/persistent/tiaf.profile.json'...

[2021-07-08T14:19:04.211Z] Is using test impact analysis: 'True'.

[2021-07-08T14:19:04.211Z] The configuration file was parsed successfully.

[2021-07-08T14:19:04.211Z] Previous commit hash found at 'D:/workspace/o3de/build/windows_vs2019/bin/TestImpactFramework/Persistent/historic\last_run.hash'.

[2021-07-08T14:19:04.470Z] Generated diff between commits '81f9afdf13c2361e09691cd85ae2ef62ed017786' and 'ea0e38aaf4b58d7142e2033d9daeeef9fd5fb58b': 'D:/workspace/o3de/build/windows_vs2019/bin/TestImpactFramework/Temp\changelist.diff'.

[2021-07-08T14:19:04.730Z] Change list constructed successfully: 'D:/workspace/o3de/build/windows_vs2019/bin/TestImpactFramework/Temp\changelist.json'.

[2021-07-08T14:19:04.730Z] 17 created files, 202 updated files and 203 deleted files.

[2021-07-08T14:19:04.730Z] Test suite is set to 'main'.

[2021-07-08T14:19:04.730Z] Test impact analysis is enabled.

[2021-07-08T14:19:04.730Z] Change list is set to 'D:/workspace/o3de/build/windows_vs2019/bin/TestImpactFramework/Temp\changelist.json'.

[2021-07-08T14:19:04.730Z] Sequence type is set to 'tianowrite'.

[2021-07-08T14:19:04.730Z] Safe mode set to 'off'.

[2021-07-08T14:19:04.730Z] Test failure policy is set to 'continue'.

[2021-07-08T14:19:04.730Z] Args: --suite=main --changelist=D:/workspace/o3de/build/windows_vs2019/bin/TestImpactFramework/Temp\changelist.json --sequence=tianowrite --safemode=off --fpolicy=continue

[2021-07-08T14:19:06.639Z] Constructing in-memory model of source tree and test coverage for test suite main, this may take a moment...

[2021-07-08T14:19:06.639Z] Test impact analysis data for this repository was found.

[2021-07-08T14:19:06.639Z] File Creation: 

[2021-07-08T14:19:06.639Z] ==================================================================

[2021-07-08T14:19:06.639Z] File Creation: Trace::Error

[2021-07-08T14:19:06.639Z]  D:/workspace/o3de/Code/Tools/TestImpactFramework/Runtime/Code/Source/Dependency/TestImpactDynamicDependencyMap.cpp(370): 'class TestImpact::ChangeDependencyList __cdecl TestImpact::DynamicDependencyMap::ApplyAndResoveChangeList(const struct TestImpact::ChangeList &)'

[2021-07-08T14:19:06.639Z] File Creation: The newly-created file Gems\ScriptCanvas\Code\Builder\ScriptCanvasBuilder.h belongs to a build target yet still has coverage data in the source covering test list implying that a delete CRUD operation has been missed, thus the integrity of the source covering test list has been compromised

[2021-07-08T14:19:06.639Z] File Creation: ==================================================================

[2021-07-08T14:19:06.639Z] The newly-created file Gems\ScriptCanvas\Code\Builder\ScriptCanvasBuilder.h belongs to a build target yet still has coverage data in the source covering test list implying that a delete CRUD operation has been missed, thus the integrity of the source covering test list has been compromised

[2021-07-08T14:19:06.639Z] The test impact analysis runtime returned with error: '5'.
```

### CODE REVIEW

File: Code/Tools/TestImpactFramework/Runtime/Code/Source/Dependency/TestImpactDynamicDependencyMap.cpp [hunk 1]
```diff
@@ -418,18 +423,26 @@ namespace TestImpact
             {
                 if (sourceDependency->GetNumCoveringTestTargets())
                 {
-                    const AZStd::string msg = AZStd::string::format("The deleted file %s still belongs to a build target and still "
+                    const AZStd::string msg = AZStd::string::format("The deleted file '%s' still belongs to a build target and still "
```
Comment (amzn-sean) [inline Code/Tools/TestImpactFramework/Runtime/Code/Source/Dependency/TestImpactDynamicDependencyMap.cpp:426]: Should this have double single quotes?

File: Code/Tools/TestImpactFramework/Runtime/Code/Source/Dependency/TestImpactDynamicDependencyMap.cpp [hunk 2]
```diff
@@ -393,7 +398,7 @@ namespace TestImpact
                     if (sourceDependency->GetNumCoveringTestTargets())
                     {
                         AZ_Printf(
-                            "File Update", AZStd::string::format("Source file '%s' is potentially an orphan (used by build targets "
+                            "File Update", AZStd::string::format("Source file ''%s'' is potentially an orphan (used by build targets "
```
Comment (amzn-sean) [inline Code/Tools/TestImpactFramework/Runtime/Code/Source/Dependency/TestImpactDynamicDependencyMap.cpp:401]: Should this have double single quotes?

**Review by amzn-sean — APPROVED** [review]
Lgtm

File: Code/Tools/TestImpactFramework/Runtime/Code/Source/Dependency/TestImpactDynamicDependencyMap.cpp (hunk 2 above)
Comment (jonawals) [reply Code/Tools/TestImpactFramework/Runtime/Code/Source/Dependency/TestImpactDynamicDependencyMap.cpp:401]: fixed 👍

**Review by SergeyAMZN — APPROVED** [review]

**Review by lumberyard-employee-dm — APPROVED** [review]

**Pull request merged**